sql/core.go

```diff
@@ -429,22 +439,42 @@ type IndexAlterableTable interface {
 	RenameIndex(ctx *Context, fromIndexName string, toIndexName string) error
 }
 
-// ForeignKeyTable is a table that can declare its foreign key constraints.
+// ForeignKeyTable is a table that can declare its foreign key constraints, as well as be referenced.
 type ForeignKeyTable interface {
-	Table
-	// GetForeignKeys returns the foreign key constraints on this table.
-	GetForeignKeys(ctx *Context) ([]ForeignKeyConstraint, error)
-}
-
-// ForeignKeyAlterableTable represents a table that supports foreign key modification operations.
-type ForeignKeyAlterableTable interface {
-	Table
-	// CreateForeignKey creates an index for this table, using the provided parameters.
-	// Returns an error if the foreign key name already exists.
-	CreateForeignKey(ctx *Context, fkName string, columns []string, referencedTable string, referencedColumns []string,
-		onUpdate, onDelete ForeignKeyReferenceOption) error
-	// DropForeignKey removes a foreign key from the database.
+	IndexedTable
+	// CreateIndexForForeignKey creates an index for this table, using the provided parameters. Indexes created through
+	// this function are specifically ones generated for use with a foreign key. Returns an error if the index name
+	// already exists, or an index on the same columns already exists.
+	CreateIndexForForeignKey(ctx *Context, indexName string, using IndexUsing, constraint IndexConstraint, columns []IndexColumn) error
+
+	// GetDeclaredForeignKeys returns the foreign key constraints that are declared by this table.
+	GetDeclaredForeignKeys(ctx *Context) ([]ForeignKeyConstraint, error)
+	// GetReferencedForeignKeys returns the foreign key constraints that are referenced by this table.
+	GetReferencedForeignKeys(ctx *Context) ([]ForeignKeyConstraint, error)
+	// AddForeignKey adds the given foreign key constraint to the table. Returns an error if the foreign key name
+	// already exists on any other table within the database.
+	AddForeignKey(ctx *Context, fk ForeignKeyConstraint) error
+	// DropForeignKey removes a foreign key from the table.
 	DropForeignKey(ctx *Context, fkName string) error
+	// SetForeignKeyResolved is called once a foreign key has been resolved.
+	SetForeignKeyResolved(ctx *Context, fkName string) error
+	// GetForeignKeyUpdater returns a ForeignKeyUpdater for this table.
+	GetForeignKeyUpdater(ctx *Context) ForeignKeyUpdater
+}
+
+// ForeignKeyUpdater is a TableEditor that allows for efficient access on indexed point lookups.
+type ForeignKeyUpdater interface {
```

Comment:
At first this was unexpected because I thought we would use the existing index lookup machinery. Then I realized that might be challenging if we need/have to return un-committed edits from the table editors. Am I thinking about this the right way? If so, is it possible to implement index lookups over un-committed edits? The new index format supports range/index queries over un-committed edits. 
I would definitely prefer to use the existing index lookup tool if possible. In the case where we're cascading a large number of rows, the current API of `GetIndexedRows` requires us to return all the matching rows in memory, rather that accessing them with a `RowIter`

Reply:
Indeed you are thinking about this in the right way.

Also a fair point on the row matching issues with the current implementation. I definitely can use a `RowIter` instead. Regarding the use of the existing index interfaces, I was originally going to use them, but they're implemented over tables rather than table editors, and I decided to implement counterparts to our current point lookup functions. I wasn't aware that range queries work on table editors which would change quite a bit, so we'll talk more about this out-of-band.